src/Swarm/TUI/Controller.hs

```diff
@@ -120,6 +125,21 @@ handleEvent s ev = do
 setFocus :: AppState -> Name -> EventM Name (Next AppState)
 setFocus s name = continue $ s & uiState . uiFocusRing %~ focusSetCurrent name
 
+toggleModal :: AppState -> Modal -> EventM Name (Next AppState)
+toggleModal s modal =
+  continue $ s & case s ^. uiState . uiModal of
+    Nothing -> (uiState . uiModal ?~ modal)   . ensurePause
+    Just _  -> (uiState . uiModal .~ Nothing) . maybeUnpause
+  where
+    -- Set the game to AutoPause if needed
+    ensurePause
```

Comment:
Hmm, actually, this doesn't quite work yet.  Notice in `handleWorldEvent` when we process a manual pause, we also reset the `lastFrameTime`, which is necessary since otherwise, once the game is unpaused, it thinks the last frame took a really long time and that it needs to catch up by doing a bunch of ticks all at once.  When we `maybeUnpause` here we don't do that, so after hitting F1, waiting for a little while, then hitting F1 again you can see the game jump ahead to where it would have been if it hadn't been paused the whole time.

I wonder if there is a good way to encapsulate this functionality so we don't have to remember to do this in every single place in the code that might want to unpause.

Reply:
Oops, good catch. I don't know how to properly encode this functionality... How about we use a `Lens' AppState -> RunStatus` which reset the lastFrameTime to 0, and in runFrameUI we could consider such value as curTime (and skip the catch up phase)?

Reply:
Hmm, but I'm not sure we want to reset the `lastFrameTime` every time we look up the `RunStatus`.  Maybe for now let's just manually reset the `lastFrameTime` in `maybeUnpause` and defer the question to later of whether we can refactor it in a better way.